### PR TITLE
which-key for stumpwm

### PR DESCRIPTION
**Pull request by dmh43**
I started using stumpWM recently but was having some trouble remembering all the keybindings. I wrote something simple that is similar to `which-key-mode` for emacs. I figured I would share. This is my first PR here, so if there is anything you'd like me to add or change, let me know.

### CODE REVIEW

**Comment by dbjergaard** [comment]
Hi,  how is this different than the built-in "where-is" ? If it does more or is a better implementation that is more true to emacs, then I would prefer that we merge the changes directly instead of as a contrib module.
Thanks for your contribution though, I look forward to more!

File: util/which-key/README.txt [hunk 1]
```diff
@@ -0,0 +1,9 @@
+This module enables functionality similar to `which-key-mode` in
+emacs.  When a keymap is enabled by pressing a key, a list of possible
+keysequences are shown to the user, similar to the help menu (`C-t ?`)
+For example, if I press `C-t` the commands in the root-map are shown
```
Comment (PuercoPop) [inline util/which-key/README.txt]: The first person should not used in documentation. Plus you are already using third person in the previous sentence. I prefer you used second person but either second or third person are fine

ej.
```
For example, when we press C-t
For example, when C-t is pressed
```

File: util/which-key/which-key.lisp [hunk 2]
```diff
@@ -0,0 +1,45 @@
+;;;; which-key.lisp
+
+(in-package #:which-key)
+
+(export '(get-kmaps-at-key-seq key-press-hook replace-hook))
```
Comment (PuercoPop) [inline util/which-key/which-key.lisp]: Although this idiom is used throughout the StumpWM codebase, it is an archaism in CL. Consider moving this to the export clause in the package definition

ej.
```lips
(defpackage #:which-key
    (:use #:cl)
    (:export #:get-kmaps-at-key
                 #:key-press-hook
                 ...)
```

File: util/which-key/which-key.lisp [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+   (reduce
+    (lambda (result map)
+      (let* ((binding (find key (kmap-bindings map)
+                            :key 'binding-key :test 'equalp))
+             (command (when binding (binding-command binding))))
+        (if command
+            (setf result (cons command result))
+            result)))
+    kmaps
+    :initial-value '())))
```
Comment (PuercoPop) [inline util/which-key/which-key.lisp]: No need to quote nil, `:initial-value ()...` would suffice

File: util/which-key/which-key.lisp [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+(defun key-press-hook (key key-seq cmd)
+  "which-key replacement for *key-press-hook*"
+  (declare (ignore key))
+  (unless (eq *top-map* *resize-map*)
+    (let ((maps (get-kmaps-at-key-seq (dereference-kmaps (top-maps))
+                                        (reverse key-seq))))
+      (when (remove-if-not 'kmap-p maps)
+        (apply 'display-bindings-for-keymaps (reverse (cdr key-seq)) maps)))))
+
+(defmacro replace-hook (hook fn)
```
Comment (PuercoPop) [inline util/which-key/which-key.lisp]: This macro only expands to `add-hook`, the `remove-hook` form is discarded as it is not returned by the macro. If you want to sequence them you have to wrap them in a `progn` form. Furthermore you are evaluating the arguments of the macro multiple times. You can work around this by using `alexandria:once-only` or writting the gensyms yourself. I prefer using alexandria but you may not want to add a dependency, if so go with the vanilla version

### Vanilla CL

```lisp
(defmacro replace-hook (hook fn)
  "Easily replace a hook with a new function"
  (let ((!fn (gensym))
        (!hook (gensym)))
    `(let ((,!fn ,fn)
           (,!hook ,hook))
       (progn (remove-hook ,!hook ,!fn)
              (add-hook ,!hook ,!fn)))))
```

### Using Alexandria
```lisp
(defmacro replace-hook (hook fn)
  "Easily replace a hook with a new function"
  (alexandria:once-only (fn hook)
    `(progn (remove-hook ,!hook ,!fn)
            (add-hook ,!hook ,!fn))))
```

**Comment by PuercoPop** [comment]
Nevermind, as @dbjergaard pointed out StumpWM already comes with `where-is`

**Comment by dbjergaard** [comment]
Is replace-hook generically useful? I don't see a case where you would run replace-hook in stumpwm unless you were running interactively.  

What I would like to do with this is: rename where-is to which-key since its more in line with emacs and it does what it says (where-is what?).  If there are other parts (replace-hook) lets put them in the right part of stumpwm for everyone. 

I think the content addressed by this module would be better served if it were applied to the code-base since its tweaking internal behavior rather than extending stumpwm's functionality.

**Comment by dmh43** [comment]
The difference between `which-key` and `where-is` is that instead of giving stumpWM a key-sequence corresponding to a command, I give it a keysequence corresponding to a keymap. It is very much like an incremental `C-t C-h`.

But yes, I agree, this probably belongs somewhere in the codebase and can take advantage of some other built in functions. I'll try to spend some time tonight on this.

Thank you for the feedback!

**Pull request closed without merging**